### PR TITLE
Add TypeScript declaration file.

### PR DESCRIPTION
**Pull request by janhesters**
## Motivation
As discussed in [this issue](https://github.com/react-navigation/react-navigation-core/issues/2) we should manage the TypeScript typings from within react-navigation.

This file extends the old DefinitelyTyped file:

What I added (and needs to be checked / reviewed): `unmountInactiveRoutes`, `drawerType` and missing `contentOptions` for `DrawerItems`.
Added `defaultNavigationOptions`.
Added default params inside route definitions.
Added `headerBackgroundTransitionPreset`.
Added `cardOverlayEnabled` and `cardShadowEnabled`.
Added `createAppContainer` (thanks to @BrendonSled).
Added `Context` and types in `package.json` (thanks to @bySabi).
 
What I couldn't add:
The support for React Hooks (I didn't find the code in V3).
~~Type definitions for `createAppContainer` (I didn't have time to look into it and understand the code).~~

**TODO:**
1. Changelog
2. React Hooks
3. See [this comment](https://github.com/react-navigation/react-navigation/pull/5276#issuecomment-444816485) below for how you can help.

## Changelog

**TODO:** Add an entry under the "Unreleased" heading in [CHANGELOG.md](https://github.com/react-navigation/react-navigation/blob/master/CHANGELOG.md#unreleased) which explains your change.

### CODE REVIEW

**Comment by bySabi** [comment]
@janhesters React Hooks support are in a separated package https://github.com/react-navigation/react-navigation-hooks. 
Maybe rewrite it with typescript will be better than create a `.d.ts` file.
I will try to contrib on this but my typescript is very limited at present time.

**Comment by jhalborg** [comment]
@bySabi - Rewriting to TS is a much larger task, even if it is the way to move forward. A proper declaration file that is maintained along with this repo instead of separately by the community on DefinitlyTyped is a great start, as it removes the "lag" there is when new updates come out

**Comment by bySabi** [comment]
My fault, sorry! .. I refer to React Hook package only, https://github.com/react-navigation/react-navigation-hooks,  not the whole `react-navigation`

File: src/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,1318 @@
...
+  removeClippedSubviews?: boolean;
+  initialLayout?: { height: number; width: number };
+}
+
+// From navigators/TabNavigator.js
+export function TabNavigator(
+  routeConfigMap: NavigationRouteConfigMap,
+  drawConfig?: TabNavigatorConfig
+): NavigationContainer;
+
```
Comment (BrendonSled) [inline src/index.d.ts]:
```suggestion
export function createAppContainer(
  routeConfigMap: NavigationRouteConfigMap,
): NavigationContainer;

```

**Comment by janhesters** [comment]
@Sharlaan It's been added today in this PR 😊

**Comment by Sharlaan** [comment]
i saw but is not merged ::)

File: src/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,1321 @@
...
+}
+export interface NavigationBottomTabScreenOptions
+  extends NavigationTabScreenOptionsBase {
+  tabBarOnPress?: (
+    options: {
+      navigation: NavigationScreenProp<NavigationRoute>;
+      defaultHandler: () => void;
+    }
+  ) => void;
```
Comment (BrendonSled) [inline src/index.d.ts]:
```suggestion
  ) => void;
    tabBarOnLongPress?: (
    options: {
      previousScene: TabScene;
      scene: TabScene;
      jumpToIndex: (index: number) => void;
    }
  ) => void;
```

**Comment by janhesters** [comment]
@BrendonSled Thank you!

**Comment by janhesters** [comment]
@brentvatne Will this be merged? Or is there anything missing? 😊

**Comment by brentvatne** [comment]
@janhesters - I would like to merge this but what makes me a little bit hesitant is that it will go out of date easily without us having some automated checks in place. For example, we used to have a precommit hook that ran flow on the NavigationPlayground project to do a quick sanity check. Ideally we would change the project to use TypeScript and run tsc on it in precommit / in CI

**Comment by janhesters** [comment]
@brentvatne Sorry english is not my first language, so just asking for clarification:

Can I rewrite React Navigation completely to TypeScript? No, as this would take way too much time. That should probably be a group effort (I am willing to help 😊).

If you mean can I rewrite the NavigationPlayground to use TypeScript, yes of course I can try. Kindly point me to it and I will get to work! 😊

**Comment by brentvatne** [comment]
> If you mean can I rewrite the NavigationPlayground to use TypeScript, yes of course I can try. Kindly point me to it and I will get to work! 😊

yup, this one! :) https://github.com/react-navigation/react-navigation/tree/master/examples/NavigationPlayground

**Comment by janhesters** [comment]
@brentvatne Took some time but I did ~80% of the migration work. The Playground is now in TypeScript. For personal reasons I can't put more time into this, but this should be a major start. Someone just has to review / finish it.

Here are the steps that still have to be done:

1. Some people who are experts on TS (I'm only using it since 4 months) should look over the declarations file and the PlayGround and review it.
2. I don't know how to use Expo and have never used is. The new Playground is without Expo. If it needs Expo, some expert on Expo has to add that.
3. The following types have just been lazily `any`ed by me, because I didn't have time to properly type them:
```
export interface HeaderStyleInterPolator {
  forLayout: any;
  forLeft: any;
  forLeftButton: any;
  forLeftLabel: any;
  forCenterFromLeft: any;
  forCenter: any;
  forRight: any;
  forBackground: any;
  forBackgroundWithInactiveHidden: any;
  forBackgroundWithFade: any;
  forBackgroundWithTranslation: any;
}

export const getActiveChildNavigationOptions: any;
```
They have to be changed in the declaration file.
4. There are four routes missing from the PlayGround right now that I couldn't migrate `CustomTranisitioner`, `TabsWithNavigationFocus`, `TabsWithNavigationEvents` and `KeyboardHandlingExample`. They still have to be migrated and rewritten with TS.

Hope this helps!

**Comment by jonbretman** [comment]
@janhesters @brentvatne I can have a go at picking this up - getting the typescript definitions published as part of this package would be a huge win for us so worth the time.

**Comment by kcolton** [comment]
### OMG THANK YOU
**Thank you all who have maintained the existing DefinitelyTyped typings and for prioritizing types in general! 🙇 🙇 ** This is _not_ an easy package to type, but they have been soooo helpful in development and the type declaration file nearly a pinned tab in IDE

some thoughts:

### maintenance:
@brentvatne have some experience w/ maintaining type declaration files and watching the cycles of drift and correction for a bunch of packages. As long as there is a decent community of people using typescript who are active then it usually winds up working out pretty well. You can always fannegal the types a bit if a little off too. 

A PR into this repo (as well as tracking if work is already in progress) is going to be more approachable than into DefinitelyTyped. 

Perhaps a [pull request template](https://help.github.com/articles/creating-a-pull-request-template-for-your-repository/) with a checkboxes like:

- [ ] Updated type definitions? (if applicable)
- [ ] Updated examples/playground? (if applicable)

as reminders for contributors to update types and examples if they are changing interfaces. 

Getting full coverage for react-navigation seems like it could be an NP-complete problem :P But the best practices from DefinitelyTyped are a great start and should be easily runnable for people and CI alike.

*examples/playground:*
idk much about the examples and playground setup yet. For someone new to this repo and setup, having the examples included with core typings change makes the diff hard to digest (low signal to noise IMHO) and intimidating (130+ files! I thought I misread and there was a TS rewrite or something :P). Maybe because it's also first-time setup code? 

Def not saying to 🗑 the examples; think they are very valuable. Thoughts on moving into a separate branch/pr or something to separate the core change and supporting materials? Just a thought from the peanut gallery - feel free to ignore

@janhesters Pretty deeply familiar with TS and react-navigation - I'll help build off the great "start" (looks closer to a finish than a start :P)

**Comment by brentvatne** [comment]
hi all! @kcolton - one concern i have with updating the issue template like that is that we probably need to do it on all react-navigation repos - eg: react-navigation-stack. the react-navigation package itself just brings together a bunch of repos these days. i'm not sure what a good workflow is for this, open to ideas. happy to merge this pr if we fix the conflicting files, looking forward to a typescript future

**Comment by alamothe** [comment]
For those of us who are super anxious (like me), here is how you can make use of @janhesters awesome typings today.

1. Get raw `index.d.ts` file from the commit. Link to download is [here](https://raw.githubusercontent.com/react-navigation/react-navigation/a84fe5615fbf9e804eaa7970daf026833ce5ade9/src/index.d.ts).
2. Make a new file inside your project named `react-navigation.d.ts`.
3. Put this inside it:
```
declare module 'react-navigation' {
  // Copy/paste downloaded file here
  //
  // You will also need to fix two errors (remove "declare" from lines 99 and 100
}
```
4. Profit! 🥇

**Comment by janhesters** [comment]
@alamothe Let me know if while using it you find any type errors so we can fix them (, or if you can suggest changes to this PR with the fixes).

File: src/index.d.ts [hunk 3]
```diff
@@ -984,6 +984,9 @@ export function TabNavigator(
   routeConfigMap: NavigationRouteConfigMap,
   drawConfig?: TabNavigatorConfig
 ): NavigationContainer;
+export function createAppContainer(
+  routeConfigMap: NavigationRouteConfigMap,
```
Comment (alamothe) [inline src/index.d.ts]: Shouldn't it take a single component (of type `NavigationContainer`?). As far as I understand, `createAppContainer` cannot take a route config map.

**Comment by alamothe** [comment]
DefinitelyTyped had a couple of commits, and the version got bumped to `3.0.0`. Please merge latest changes (it has properly typed `createAppContainer`).

**Comment by janhesters** [comment]
@alamothe Done

**Comment by timsawtell-sportsbet** [comment]
Anything I can do to help this one along? Eagerly waiting this PR before we upgrade to v3 :)

File: src/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,1373 @@
...
+  containerConfig?: any;
+  contentOptions?: {
+    activeTintColor?: string;
+    activeBackgroundColor?: string;
+    inactiveTintColor?: string;
+    inactiveBackgroundColor?: string;
+    style?: StyleProp<ViewStyle>;
+    labelStyle?: StyleProp<TextStyle>;
+  };
+  drawerLockMode?: DrawerLockMode;
```
Comment (koenpunt) [inline src/index.d.ts]: `drawerType` and `overlayColor` are missing here:

```ts
drawerType?: 'front' | 'back' | 'slide';
overlayColor?: string;
```

File: src/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,1373 @@
...
+/**
+ * Drawer Navigator
+ */
+export interface DrawerViewConfig {
+  drawerBackgroundColor?: string;
+  drawerWidth?: number;
+  drawerPosition?: 'left' | 'right';
+  contentComponent?: React.ComponentType<DrawerItemsProps>;
+  contentOptions?: any;
+  style?: StyleProp<ViewStyle>;
```
Comment (koenpunt) [inline src/index.d.ts]: `drawerType` and `overlayColor` are missing here, and `drawerWidth` also accepts a function;

```ts
drawerType?: 'front' | 'back' | 'slide';
overlayColor?: string;
drawerWidth?: number | (() => number);
```

**Comment by lucsky** [comment]
Is it me or are wrapped scrollables (see https://reactnavigation.org/docs/en/scrollables.html) missing from the types definition ?

**Comment by bySabi** [comment]
@lucsky your are right is not any type on Scrollables on current Typescript or Flow definitions. 

Scrollables come from `react-navigation-native` https://github.com/react-navigation/react-navigation-native/blob/master/src/Scrollables.js . I'm porting this module to TS and all the types there will be exported. 
I have ported `react-navigation@core` to TS too but is no yet merged or reviewed. Any insight will be welcome: https://github.com/bySabi/react-navigation-core/tree/move_to_typescript

**Comment by bySabi** [comment]
@alamothe @timsawtell-sportsbet @koenpunt I hope that https://github.com/bySabi/react-navigation-core/tree/move_to_typescript branch will be the future base of typings for `react-navigation`
 
Any help reviewing the code will be great.

**Comment by brentvatne** [comment]
i'm going to have to convert the navigationplayground app back to using expo as i keep it in expo in order to be able to publish it so users can open it from the expo client in ios simulator and on android device/simulator, if someone wants to do this it'd help speed up merging this

**Comment by jimrejnor** [comment]
I have installed latest version of @types/react-navigation (3.0.1), but I can't seem to find `getNavigation` method. I'm working on MobX navigation state management so it would be nice to have @types for that method.

**Comment by sinewave440hz** [comment]
Um...is anything happening with this at the mo? I just tried 3.0.4 and still getting missing drawerType as mentioned above.

**Comment by brentvatne** [comment]
i don't have time to work on it at the moment, you can help with it if you want @sinewave440hz

**Comment by brentvatne** [comment]
hey all, I landed this in https://github.com/react-navigation/react-navigation/commit/eac5feb64e2170f370a815f10af77b471e146648 but it's not ready to be published yet probably. please try it out and submit PR to improve the types! thanks everyone for the work on this!

**Pull request closed without merging**